plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,565 @@
+            this.s3AsyncClient = requireNonNull(s3AsyncClient, "s3AsyncClient is null");
+            this.bucketName = requireNonNull(bucketName, "bucketName is null");
+            this.key = requireNonNull(key, "key is null");
+            this.partSize = partSize;
+            this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        }
+
+        @Override
+        public ListenableFuture<Void> write(Slice slice)
+                throws IOException
+        {
+            // skip multipart upload if there would only be one part
+            if (slice.length() < partSize && uploadId.isEmpty()) {
+                PutObjectRequest.Builder putObjectRequestBuilder = PutObjectRequest.builder()
+                        .bucket(bucketName)
+                        .key(key);
+                S3RequestUtil.configureEncryption(secretKey, putObjectRequestBuilder);
+
+                try {
+                    s3Client.putObject(putObjectRequestBuilder.build(),
+                            // avoid extra memory copy
+                            fromContentProvider(() -> new ByteArrayInputStream(slice.getBytes()), slice.length(), Mimetype.MIMETYPE_OCTET_STREAM));
+                    return immediateVoidFuture();
+                }
+                catch (AwsServiceException e) {
+                    throw new IOException(e);
+                }
+            }
+
+            if (uploadId.isEmpty()) {
+                uploadId = Optional.of(createMultipartUpload().uploadId());
+            }
+
+            CompletableFuture<CompletedPart> uploadFuture = uploadPart(uploadId.get(), slice);
+            uploadFutures.add(uploadFuture);
+
+            return asVoid(toListenableFuture(uploadFuture));
+        }
+
+        @Override
+        public void close()
+                throws IOException
+        {
+            if (uploadId.isEmpty()) {
+                return;
+            }
+
+            try {
+                List<CompletedPart> completedParts = uploadFutures.stream()
+                        .map(CompletableFuture::join)
+                        .sorted(Comparator.comparing(CompletedPart::partNumber))
+                        .collect(toImmutableList());
+                completeMultiUpload(uploadId.get(), completedParts);
+            }
+            catch (RuntimeException e) {
+                abortUploadSuppressed(uploadId.get(), e);
+                throw new IOException(e);
+            }
+        }
+
+        private CreateMultipartUploadResponse createMultipartUpload()
+        {
+            CreateMultipartUploadRequest.Builder createMultipartUploadRequestBuilder = CreateMultipartUploadRequest.builder()
+                    .bucket(bucketName)
+                    .key(key);
+            S3RequestUtil.configureEncryption(secretKey, createMultipartUploadRequestBuilder);
+            return s3Client.createMultipartUpload(createMultipartUploadRequestBuilder.build());
```

Comment:
This is currently blocking. I wonder how difficult is it to implement it in a non blocking way?

Reply:
This is a straightforward one, will address it.

Reply:
`CompletableFuture`'s `thenCompose` will run the later stage in the same thread. `thenComposeAsync` will run the later stage in the default thread pool (instead of AWS client's Netty thread pool). Feel in this case it makes less sense to do `createMultipartUpload` in a non-blocking way.

Please advise.